### ISSUE TITLE
Envoy ignores H/2 max concurrent streams advertised by peers

### ISSUE THREAD

**Issue by kryzthov-stripe** [envoyproxy/envoy#13933]
*Title*: Envoy ignores H/2 max concurrent streams advertised by peers

*Description*:
Envoy does not seem to correctly calculate the proper max concurrent streams to use.
Envoy seems to always use the max concurrent streams from its own configuration, and disregards the max concurrent streams advertised by a backend through the H/2 settings frames.
This can lead to situations where Envoy does not create new H/2 connections, and submits requests through an existing nghttp2 session that is currently at or above stream capacity.
nghttp2 handles this "gracefully" by queuing the requests internally until stream capacity opens again.
Envoy will count such requests as active (ie. will count against upstream_rq_active), despite the requests being queued in the nghttp2 session and not being actually sent to the upstream.

*Repro steps*:
This can be reproduced by standing up an H/2 backend advertising low max concurrent streams settings:
https://gist.github.com/kryzthov-stripe/8d8b5f5f1bd7cb99a88baba08f1c5058
And fronting this backend with Envoy with a high max concurrent streams settings (the default here is 2 billions):
https://gist.github.com/kryzthov-stripe/95e2309344113b7a282ffc7dbfc80d86
To make the problem more obvious, restrict Envoy's concurrent to 1:
```
envoy --config-path envoy.yaml --concurrency 1
```
Then sending concurrent requests with something like hey (`go get github.com/rakyll/hey`):
```
hey -c 200 -t 0 http://localhost:10000
```

In the example above, Envoy will only establish a single H/2 connection to the backend (single worker thread), and direct all concurrent requests to the connection. The backend restricts the number of concurrent streams to 5, which nghttp2 session will respect by queuing the excess streams internally. The backend logs confirm that only 5 requests are being processed at a time.

I would expect Envoy to recognize that the backend can only process 5 streams concurrently, and reconfigure the client accordingly. This should lead Envoy to create new H/2 connections as needed and permitted by circuit breakers. In the example above, I would expect Envoy to create ~40 connections.

**Comment by ggreenway** [comment]
Prior to early 2020, Envoy's http/2 connection pool could only have 1 active upstream connection. This has been fixed to support many connections, but it was never fixed to respect the peer settings frame. I don't think this would be too hard to implement. A few things I can see that would need doing (this list may be incompletey):

* In `source/common/http/http2/codec_impl.cc`, in `ConnectionImpl::onFrameReceived(const nghttp2_frame* frame)`, note when a SETTINGS frame is received and get the value for max concurrent streams, if present.
* Pass this value to the `ActiveClient` in `source/common/http/http2/conn_pool.cc`. A value from the config `max_concurrent_streams` is currently passed to the base class constructor. Update this value when the SETTINGS frame arrives.
* Figure out if anything needs to be done if the new value is less than the current number of streams (which can happen if a bunch of streams are established immediately after the connection is setup, before the peer SETTINGS are received).

**Comment by ggreenway** [comment]
Also, consider removing the following, and making the `ActiveClient` use an initial value of 100.

https://github.com/envoyproxy/envoy/blob/master/source/common/http/http2/codec_impl.cc#L1376
```
  // Temporarily disable initial max streams limit/protection, since we might want to create
  // more than 100 streams before receiving the HTTP/2 SETTINGS frame from the server.
  //
  // TODO(PiotrSikora): remove this once multiple upstream connections or queuing are implemented.
  nghttp2_option_set_peer_max_concurrent_streams(
      options_, ::Envoy::Http2::Utility::OptionsLimits::DEFAULT_MAX_CONCURRENT_STREAMS);
```

**Comment by antoniovicente** [comment]
cc @alyssawilk 

Relying on internal nghttp2 buffering for this case doesn't seem great.  I think that a possible source of complexity is if the upstream does not provide a consistent max_streams value across connections; the upstream connection selection code needs to be aware of individual connection capacities and only select connections that are below the saturation value.  This also has implications to upstream connection prefetch; I think the current (in-progress) implementation is based on values specified in the Envoy config.

**Comment by alyssawilk** [comment]
Yeah, the H2 connection pool should totally adjust max streams per connection to be min(configured, settings-negotiated).  
I'll probably be adding helper functions to adjust max stream per connections for the ALPN pool (once the prefetch and ALPN PS combine) so then it'd just be a case of intercepting SETTINGS and passing the new limit from the codec client to the connection pool.

**Comment by antoniovicente** [comment]
What if the negotiated number of streams is different for every upstream connection created to a specific endpoint?

**Comment by alyssawilk** [comment]
That's fine - the ALPN pool already adjusts the per-connection limits on connection establishment (in that case, max streams for H2, 1 for H1) so it shouldn't be different to adjust the per-connection limits for H2 to the min of (pool-wide concurrent streams cap, settings negotiated cap)

**Comment by alyssawilk** [comment]
Well the fix to this *was* simple, until I ruined it by implementing prefetching.  @mattklein123 and @antoniovicente can I get your thoughts on the least ugly solutoin?

The new wrinkle is that we track capacity across a cluster.  If we open a new TCP connection which would allow 5 new streams, we increment our stream capacity by 5.  If that connection is established and we create a stream on it, we decrement it to 4 (as we could handle 4 new incoming streams).  When that stream ends, we add one and capacity returns to 5.

Now imagine we open a new HTTP/2 connection with anticipatd capacity 100.   (capacity == 100).  We immediately ship 5 queued streams (capacity == 95).  Then we get the SETTINGS frame back saying stream limit is 1.  4 of those 5 streams will be reset.  For accounting we could
1) internal to ClusterConnectivityState track capacity as an unt64_t, and let it go down to (-4) so as those 4 streams are reset it goes to 0, and as the 5th is served it becomes 1.  capacity accessors would simply see capacity as 0.
2) Set capacity immediately to 0, but change the stream accounting to track the delta and not "give back" capacity while there is an outstanding delta (store that -4 in the HTTP2 connection pool somewhere)
3) change the HTTP/2 connection pool to not establish more than 1 stream until SETTINGS arrives (I think this is a no-go for latency reasons)

without input I'm inclined to go with 2, but I really dislike both 1 and 2 so ideals welcome.

**Comment by alyssawilk** [comment]
https://github.com/envoyproxy/envoy/compare/master...alyssawilk:stream_limit?expand=1 (Search for FIXME) if code pointers are helpful

**Comment by antoniovicente** [comment]
Waiting until Server SETTINGS arrives before declaring the H2 connection established seems like a good option to me.

Assuming 1 stream until after SETTINGS exhange could result in multiple connections being established to the same host and then quickly closed when the proxy realizes that the stream limit is high enough and a smaller number of upstream connections are needed.  Assuming that the upstream H2 connection will be able to handle 5-10 streams (overridable by config) would avoid creating unneeded connections.

Limiting H2 connections to a single stream until settings arrive doesn't seem too bad if some prefetching is in use since prefetching should hide the connection establishment times.  The exception being just after the cluster/endpoint gets the first few requests and the proxy happens to have more than 1 queued request for an endpoint.

Allowing the capacity to go negative by storing it in an int64_t but returning max(0, capacity) from capacity accessors seems reasonable.

**Comment by alyssawilk** [comment]
Even if we were willing to take the latency hit for the common case (which I'm not sure is true) it turns out that wouldn't solve the problem:

https://tools.ietf.org/html/rfc7540#section-6.5
A SETTINGS frame MUST be sent by both endpoints at the start of a
   connection and MAY be sent at any other time by either endpoint over
   the lifetime of the connection.

That brings us back to handling the negative somewhere (1 or 2) or relying on nghttp2 queueing for mid-connection SETTINGS :-/

**Comment by antoniovicente** [comment]
> Even if we were willing to take the latency hit for the common case (which I'm not sure is true) it turns out that wouldn't solve the problem:
> 
> https://tools.ietf.org/html/rfc7540#section-6.5
> A SETTINGS frame MUST be sent by both endpoints at the start of a
> connection and MAY be sent at any other time by either endpoint over
> the lifetime of the connection.
> 
> That brings us back to handling the negative somewhere (1 or 2) or relying on nghttp2 queueing for mid-connection SETTINGS :-/

SETTINGs being required early on is agood.

It would be good to understand how common mid-connection settings are.  Mid connection settings that reduce stream limits are likely to cause problems.  I guess servers should be flexible after reducing the limits.

**Comment by alyssawilk** [comment]
I'm not sure everyone would agree to take the latency hit when doing in-network traffic (when they set the limits on both ends)

I don't think it matters how common mid-connection settings are.  If they're allowed by the spec we have to handle them.  I doubt anyone lowers the stream limit in practice but I'm not going to have overflow code lying around, or kick back to nghttp2 buffering for legitimate if unusual.

So any opinions between 1) and 2) or should I just pick one?  :-)

**Comment by antoniovicente** [comment]
Options 1 and 2 seem equivalent from the outside.  Pick one.

**Comment by mattklein123** [comment]
> Options 1 and 2 seem equivalent from the outside. Pick one.

Agreed, both sound reasonable to me.
